src/app/qgsstatisticalsummarydockwidget.cpp

```diff
@@ -131,35 +142,74 @@ void QgsStatisticalSummaryDockWidget::refreshStatistics()
     mPreviousFieldType = mFieldType;
   }
 
+  QString sourceFieldExp = mFieldExpressionWidget->currentField();
   bool selectedOnly = mSelectedOnlyCheckBox->isChecked();
 
-  switch ( mFieldType )
+  if ( mGatherer )
   {
-    case DataType::Numeric:
-      updateNumericStatistics( selectedOnly );
-      break;
-    case DataType::String:
-      updateStringStatistics( selectedOnly );
-      break;
-    case DataType::DateTime:
-      updateDateTimeStatistics( selectedOnly );
-      break;
-    default:
-      break;
+    mGatherer->cancel();
+  }
+
+  bool ok;
+  QgsFeatureIterator fit = QgsVectorLayerUtils::getValuesIterator( mLayer, sourceFieldExp, ok, selectedOnly );
+  if ( ok )
+  {
+    int featureCount = selectedOnly ? mLayer->selectedFeatureCount() : mLayer->featureCount();
+    mGatherer = new QgsStatisticsValueGatherer( mLayer, fit, featureCount, sourceFieldExp );
+    switch ( mFieldType )
+    {
+      case DataType::Numeric:
+        connect( mGatherer, &QgsStatisticsValueGatherer::taskCompleted, this, &QgsStatisticalSummaryDockWidget::updateNumericStatistics );
+        break;
+      case DataType::String:
+        connect( mGatherer, &QgsStatisticsValueGatherer::taskCompleted, this, &QgsStatisticalSummaryDockWidget::updateStringStatistics );
+        break;
+      case DataType::DateTime:
+        connect( mGatherer, &QgsStatisticsValueGatherer::taskCompleted, this, &QgsStatisticalSummaryDockWidget::updateDateTimeStatistics );
+        break;
+      default:
+        break;
+    }
+    connect( mGatherer, &QgsStatisticsValueGatherer::progressChanged, mCalculatingProgressBar, &QProgressBar::setValue );
+    connect( mCancelButton, &QPushButton::clicked, mGatherer, &QgsStatisticsValueGatherer::cancel );
+    mCalculatingProgressBar->setMinimum( 0 );
+    mCalculatingProgressBar->setMaximum( featureCount > 0 ? 100 : 0 );
+    mCalculatingProgressBar->setValue( 0 );
+    mCancelButton->show();
+    mCalculatingProgressBar->show();
+
+    QgsApplication::taskManager()->addTask( mGatherer );
   }
 }
 
-void QgsStatisticalSummaryDockWidget::updateNumericStatistics( bool selectedOnly )
+void QgsStatisticalSummaryDockWidget::gathererFinished()
 {
-  QString sourceFieldExp = mFieldExpressionWidget->currentField();
+  mGatherer->deleteLater();
```

Comment:
No need for this - task manager takes ownership, so this task will be automatically deleted.